strax/storage/mongo.py

```diff
@@ -0,0 +1,137 @@
+"""I/O format for MongoDB
+
+This plugin is designed with data monitoring in mind, to put smaller
+amounds of extracted data into a database for quick access. However
+it should work with any plugin.
+
+Note that there is no check to make sure the 16MB document size
+limit is respected!
+"""
+
+try:
+    from pymongo import MongoClient
+except ImportError:
+    MongoClient = None
+
+import strax
+import numpy as np
+from strax import StorageFrontend, StorageBackend, DataKey
+from strax.utils import unpack_dtype
+from bson import json_util
+import copy
+
+export, __all__ = strax.exporter()
+
+
+@export
+class MongoBackend(StorageBackend):
+
+    def __init__(self, uri, database):
+        if MongoClient is None:
+            raise ImportError("Pymongo did not import: "
+                              "cannot use MongoDBStore")
+        self.client = MongoClient(uri)
+        self.db = self.client[database]
+
+    def _read_chunk(self, backend_key, chunk_info, dtype, compressor):
+
+        doc = self.db[backend_key].find_one(
+            {'chunk': chunk_info['chunk_i']})
+        if doc is None:
+            return []
+
+        result = np.zeros(len(doc['data']), dtype=dtype)
+        for i, d in enumerate(doc['data']):
+            for key in np.dtype(dtype).names:
+                result[key][i] = d[key]
+        return result
+
+    def _saver(self, key, metadata):
+        return MongoSaver(key, metadata, self.db[str(key)])
+
+    def get_metadata(self, key):
+        return self.db[key].find_one({'metadata': True})
+
+
+@export
+class MongoFrontend(StorageFrontend):
+    """MongoDB storage frontend"""
+
+    def __init__(self, uri, database, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if MongoClient is None:
+            raise ImportError("Pymongo did not import: "
+                              "cannot use MongoDBStore")
+        self.client = MongoClient(uri)
+        self.db = self.client[database]
+        self.backends = [MongoBackend(uri, database)]
+
+    def _find(self, key, write, allow_incomplete, fuzzy_for,
+              fuzzy_for_options):
+
+        if write:
+            return self.backends[0].__class__.__name__, str(key)
+        if self.db[str(key)].count():
+            self.log.debug(f"{key} is in cache.")
+            return self.backends[0].__class__.__name__, str(key)
+        self.log.debug(f"{key} is NOT in cache.")
+        raise strax.DataNotAvailable
+
+    def _read_meta(self, coll_name):
```

Comment:
This is not used right?

Reply:
Thanks for the suggestions. Updated.

One thing I just noticed is that I never call 'drop'. Is it standard in strax to leave partial data written on disk or clear it in case of a crash? I saw the allow_partial option so I'm assuming I could use that if I want to handle it somehow.

Reply:
Thanks! Indeed it's standard to leave it; this dates back to when we thought online deletion was necessary. The only time strax clears data is while overwriting, I guess you could drop the collection if it already exists when you init the saver, just like in https://github.com/AxFoundation/strax/blob/master/strax/storage/files.py#L277. The frontend has an allow_overwrite option that can prevent savers getting created if the data already exists.